File: src/MovieRental/Customer.cs
namespace MovieRental;

public class Customer
{
    private string _name;
    private List<Rental> _rentals = new List<Rental>();

    public Customer(string name)
    {
        _name = name;
    }

    public void AddRental(Rental rental)
    {
        _rentals.Add(rental);
    }

    public string Name => _name;

    public string Statement()
    {
        double totalAmount = 0;
        int frequentRenterPoints = 0;
        string result = "Rental Record for " + Name + "\n";

        foreach (Rental each in _rentals)
        {
            double thisAmount = 0;

            // determine amounts for each line
            switch (each.Movie.PriceCode)
            {
                case Movie.Regular:
                    thisAmount += 2;
                    if (each.DaysRented > 2)
                        thisAmount += (each.DaysRented - 2) * 1.5;
                    break;
                case Movie.NewRelease:
                    thisAmount += each.DaysRented * 3;
                    break;
                case Movie.Childrens:
                    thisAmount += 1.5;
                    if (each.DaysRented > 3)
                        thisAmount += (each.DaysRented - 3) * 1.5;
                    break;
            }

            // add frequent renter points
            frequentRenterPoints++;
            // add bonus for a two day new release rental
            if (each.Movie.PriceCode == Movie.NewRelease && each.DaysRented > 1)
                frequentRenterPoints++;

            // show figures for this rental
            result += "\t" + each.Movie.Title + "\t" + thisAmount + "\n";
            totalAmount += thisAmount;
        }

        // add footer lines
        result += "Amount owed is " + totalAmount + "\n";
        result += "You earned " + frequentRenterPoints + " frequent renter points";

        return result;
    }
}

File: src/movierental/customer.py
from movierental.movie import Movie


class Customer:

    def __init__(self, name):
        self._name = name
        self._rentals = []

    def add_rental(self, rental):
        self._rentals.append(rental)

    @property
    def name(self):
        return self._name

    def statement(self):
        total_amount = 0.0
        frequent_renter_points = 0
        result = "Rental Record for " + self.name + "\n"

        for each in self._rentals:
            this_amount = 0.0

            # determine amounts for each line
            if each.movie.price_code == Movie.REGULAR:
                this_amount += 2
                if each.days_rented > 2:
                    this_amount += (each.days_rented - 2) * 1.5
            elif each.movie.price_code == Movie.NEW_RELEASE:
                this_amount += each.days_rented * 3
            elif each.movie.price_code == Movie.CHILDRENS:
                this_amount += 1.5
                if each.days_rented > 3:
                    this_amount += (each.days_rented - 3) * 1.5

            # add frequent renter points
            frequent_renter_points += 1
            # add bonus for a two day new release rental
            if each.movie.price_code == Movie.NEW_RELEASE and each.days_rented > 1:
                frequent_renter_points += 1

            # show figures for this rental
            result += "\t" + each.movie.title + "\t" + str(this_amount) + "\n"
            total_amount += this_amount

        # add footer lines
        result += "Amount owed is " + str(total_amount) + "\n"
        result += "You earned " + str(frequent_renter_points) + " frequent renter points"

        return result
